src/target/sam3x.c

```diff
@@ -195,9 +228,152 @@ static size_t sam_flash_size(uint32_t cidr)
 	return 0;
 }
 
+static size_t sam_sram_size(uint32_t cidr)
+{
+	switch (cidr & CHIPID_CIDR_SRAMSIZ_MASK) {
```

Comment:
You shouldn't hard-code the SRAM size here. Some of these devices have dTCM and iTCM which steal from the SRAM array. The sizes are configured by GPNVM bits in the Flash controller. Some configurations permit all SRAM to be allocated to xTCM, so SRAM size in this situation will be zero.

See [Maven's SAMx7X](https://github.com/Syncena/Maven-firmware/blob/main/src/cortexm/soc_samx7x.c) for an example.

Reply:
I'm aware of the TCM, but I'm unsure if I should do this dynamic setting of ram, if a command to set  GPNVM bits is done I would need to recalculate and reset the ram, besides physically the ram is there, it just has a different purpose, depends on what exactly blackmagic or gdb does with this

@dragonmux is re-setting the ram possible? should the ram be dynamically set?

Reply:
for reference,

TCM is "Tightly Coupled Memory", essentially a region of ram that is directly accessible by the processor at it's running clock, used for fast code execution/memory access, it is dynamically set by 2 GPNVM (like fuse bits), and makes use of the SRAM

Reply:
Hm I'm thinking about this wrong, the ram would still be mapped at 0x20400000, just with a smaller size, so you're right, I should do dynamic setting, though for reference, from he datasheet `"Accesses made to the SRAM above the size limit will not generate aborts"`